File: apps/mobile/smarttaxi_app/lib/features/driver/models/navigation_progress.dart
import 'dart:math' as math;

import 'package:latlong2/latlong.dart';
import '../../shared/models.dart';

/// Distances are measured along the road, never across a block or hairpin.
class NavigationProgress {
  const NavigationProgress(
      {required this.alongMeters,
      required this.totalMeters,
      required this.distanceFromRoute,
      required this.distanceMeters,
      required this.durationSeconds});
  final double alongMeters;
  final double totalMeters;
  final double distanceFromRoute;
  final double distanceMeters;
  final double durationSeconds;
}

const _earthRadius = 6371000.0;
double _meters(LatLng a, LatLng b) {
  final lat = (b.latitude - a.latitude) * math.pi / 180;
  final lng = (b.longitude - a.longitude) * math.pi / 180;
  final h = math.pow(math.sin(lat / 2), 2) +
      math.cos(a.latitude * math.pi / 180) *
          math.cos(b.latitude * math.pi / 180) *
          math.pow(math.sin(lng / 2), 2);
  return 2 * _earthRadius * math.asin(math.sqrt(h.clamp(0.0, 1.0)));
}

({double along, double distance, double total}) _project(
    List<LatLng> line, LatLng point,
    {double? expectedAlong, double minimumAlong = 0}) {
  var total = 0.0, bestAlong = 0.0, bestDistance = double.infinity;
  final cosLat = math.cos(point.latitude * math.pi / 180);
  for (var i = 0; i < line.length - 1; i++) {
    final a = line[i], b = line[i + 1];
    final segment = _meters(a, b);
    final dx = (b.longitude - a.longitude) * cosLat;
    final dy = b.latitude - a.latitude;
    final px = (point.longitude - a.longitude) * cosLat;
    final py = point.latitude - a.latitude;
    final length2 = dx * dx + dy * dy;
    final t =
        length2 == 0 ? 0.0 : ((px * dx + py * dy) / length2).clamp(0.0, 1.0);
    final along = total + segment * t;
    final projected = LatLng(
        a.latitude + dy * t, a.longitude + (b.longitude - a.longitude) * t);
    final distance = _meters(point, projected);
    final tied = (distance - bestDistance).abs() < 0.5;
    if (along >= minimumAlong - 0.5 &&
        (distance < bestDistance - 0.5 ||
            (tied &&
                expectedAlong != null &&
                (along - expectedAlong).abs() <
                    (bestAlong - expectedAlong).abs()))) {
      bestDistance = distance;
      bestAlong = along;
    }
    total += segment;
  }
  return (along: bestAlong, distance: bestDistance, total: total);
}

NavigationProgress? navigationProgress(
    RoutePreview route, Coordinate position) {
  if (route.isFallback ||
      route.geometry.length < 2 ||
      !position.lat.isFinite ||
      !position.lng.isFinite) {
    return null;
  }
  final p = _project(route.geometry, position.toLatLng());
  // The reroute threshold is 60m. Do not continue issuing instructions from
  // the old road while waiting for its replacement.
  if (p.distance > 60 || p.total <= 0) return null;
  final fraction = ((p.total - p.along) / p.total).clamp(0.0, 1.0);
  return NavigationProgress(
      alongMeters: p.along,
      totalMeters: p.total,
      distanceFromRoute: p.distance,
      distanceMeters: route.distanceMeters * fraction,
      durationSeconds: route.durationSeconds * fraction);
}

({RouteStep step, double distanceMeters})? nextNavigationStep(
    RoutePreview route, NavigationProgress progress) {
  if (route.isFallback) return null;
  final stepTotal =
      route.steps.fold(0.0, (sum, step) => sum + step.distanceMeters);
  var traversed = 0.0, previousAnchor = 0.0;
  for (final step in route.steps) {
    final expected = step.type == 'arrive'
        ? progress.totalMeters
        : stepTotal > 0
            ? traversed / stepTotal * progress.totalMeters
            : null;
    final anchor = _project(route.geometry, step.location.toLatLng(),
        expectedAlong: expected, minimumAlong: previousAnchor);
    traversed += step.distanceMeters;
    if (!anchor.distance.isFinite || anchor.distance > 30) continue;
    previousAnchor = anchor.along;
    if (step.type == 'depart' || anchor.along < progress.alongMeters - 5) {
      continue;
    }
    return (
      step: step,
      distanceMeters: math.max(0, anchor.along - progress.alongMeters) *
          route.distanceMeters /
          progress.totalMeters
    );
  }
  return null;
}

bool navigationFixIsFresh(DateTime? timestamp, DateTime now) {
  if (timestamp == null) return false;
  final age = now.difference(timestamp);
  return age >= const Duration(seconds: -5) &&
      age <= const Duration(seconds: 12);
}

/// An immediate instruction takes precedence even if the navigator was opened
/// only 15m before the turn; never announce "in 200m" at that point.
int navigationAnnouncementStage(double distanceMeters, int previousStage) {
  if (!distanceMeters.isFinite || distanceMeters < 0) return previousStage;
  if (distanceMeters <= 40) return 2;
  if (distanceMeters <= 200 && previousStage < 1) return 1;
  return previousStage;
}

File: apps/web/src/features/driver/navigationProgress.js
// Same road-distance model as Flutter's navigation_progress.dart. No turns
// are invented from bends in a polyline or from a fallback straight line.
const radians = Math.PI / 180;
const numeric = value => typeof value === "number" && Number.isFinite(value);
const clamp = (value, low, high) => Math.min(high, Math.max(low, value));
function validPoint(point) {
  return numeric(point?.lat) && numeric(point?.lng) && Math.abs(point.lat) <= 90 && Math.abs(point.lng) <= 180;
}
function meters(a, b) {
  const h = Math.sin((b.lat - a.lat) * radians / 2) ** 2 +
    Math.cos(a.lat * radians) * Math.cos(b.lat * radians) * Math.sin((b.lng - a.lng) * radians / 2) ** 2;
  return 12742000 * Math.asin(Math.sqrt(clamp(h, 0, 1)));
}
function project(line, point, expectedAlong = null, minimumAlong = 0) {
  let total = 0, bestAlong = 0, bestDistance = Infinity;
  const cos = Math.cos(point.lat * radians);
  for (let i = 0; i < line.length - 1; i++) {
    const a = line[i], b = line[i + 1];
    const length = meters(a, b);
    const dx = (b.lng - a.lng) * cos, dy = b.lat - a.lat;
    const px = (point.lng - a.lng) * cos, py = point.lat - a.lat;
    const length2 = dx * dx + dy * dy;
    const t = length2 ? clamp((px * dx + py * dy) / length2, 0, 1) : 0;
    const along = total + length * t;
    const distance = meters(point, { lat: a.lat + dy * t, lng: a.lng + (b.lng - a.lng) * t });
    const tied = Math.abs(distance - bestDistance) < 0.5;
    if (along >= minimumAlong - 0.5 && (distance < bestDistance - 0.5 ||
        (tied && expectedAlong !== null && Math.abs(along - expectedAlong) < Math.abs(bestAlong - expectedAlong)))) {
      bestDistance = distance;
      bestAlong = along;
    }
    total += length;
  }
  return { along: bestAlong, distance: bestDistance, total };
}

export function navigationFixIsFresh(timestamp, now = Date.now()) {
  return numeric(timestamp) && now - timestamp >= -5000 && now - timestamp <= 12000;
}

export function browserNavigationFix(position, previousTimestamp = null, now = Date.now()) {
  const point = { lat: position?.coords?.latitude, lng: position?.coords?.longitude };
  const timestamp = position?.timestamp;
  if (!validPoint(point) || !navigationFixIsFresh(timestamp, now) ||
      (numeric(previousTimestamp) && timestamp <= previousTimestamp)) return null;
  const { heading, speed, accuracy } = position.coords;
  return { ...point, timestamp,
    heading: numeric(heading) && heading >= 0 && heading <= 360 ? heading % 360 : null,
    speed: numeric(speed) && speed >= 0 && speed <= 120 ? speed : null,
    accuracy: numeric(accuracy) && accuracy >= 0 ? accuracy : null
  };
}

export function navigationProgress(route, position) {
  if (!route || route.fallback || route.providerStatus === "Fallback" || !validPoint(position) ||
      !numeric(route.distanceMeters) || route.distanceMeters < 0 ||
      !numeric(route.durationSeconds) || route.durationSeconds < 0) return null;
  const coordinates = route.geometry?.type === "LineString" && route.geometry.coordinates;
  if (!Array.isArray(coordinates) || coordinates.length < 2) return null;
  const line = coordinates.map(p => Array.isArray(p) ? { lng: p[0], lat: p[1] } : null);
  if (!line.every(validPoint)) return null;
  const projection = project(line, position);
  if (projection.distance > 60 || projection.total <= 0) return null;
  const fraction = clamp((projection.total - projection.along) / projection.total, 0, 1);
  return { alongMeters: projection.along, totalMeters: projection.total,
    distanceFromRoute: projection.distance, distanceMeters: route.distanceMeters * fraction,
    durationSeconds: route.durationSeconds * fraction, line };
}

export function nextNavigationStep(route, progress) {
  if (!progress || route?.fallback || !Array.isArray(route?.steps)) return null;
  const steps = route.steps.filter(step => validPoint(step) && numeric(step.distanceMeters) && step.distanceMeters >= 0);
  const stepTotal = steps.reduce((sum, step) => sum + step.distanceMeters, 0);
  let traversed = 0, previousAnchor = 0;
  for (const step of steps) {
    const expected = step.type === "arrive" ? progress.totalMeters : stepTotal > 0 ? traversed / stepTotal * progress.totalMeters : null;
    const anchor = project(progress.line, step, expected, previousAnchor);
    traversed += step.distanceMeters;
    if (!Number.isFinite(anchor.distance) || anchor.distance > 30) continue;
    previousAnchor = anchor.along;
    if (step.type === "depart" || anchor.along < progress.alongMeters - 5) continue;
    return { step, distanceMeters: Math.max(0, anchor.along - progress.alongMeters) * route.distanceMeters / progress.totalMeters };
  }
  return null;
}

export function navigationInstruction(step) {
  if (!step) return { text: "Следуйте по маршруту", icon: "route" };
  const modifier = step.modifier;
  if (step.type === "arrive") return { text: "Точка назначения", icon: "finish" };
  if (["roundabout", "rotary", "roundabout turn"].includes(step.type)) return {
    text: Number.isInteger(step.exit) && step.exit > 0 ? `Круговое движение · ${step.exit}-й съезд` : "Круговое движение", icon: "roundabout"
  };
  if (["exit roundabout", "exit rotary"].includes(step.type)) return { text: "Съезд с кругового движения", icon: "right" };
  if (modifier === "uturn") return { text: "Развернитесь", icon: "uturn" };
  const direction = modifier?.includes("left") ? "left" : modifier?.includes("right") ? "right" : "straight";
  const side = direction === "left" ? "левее" : "правее";
  if (step.type === "fork" && direction !== "straight") return { text: `Держитесь ${side}`, icon: direction };
  if (step.type === "merge") return { text: "Перестройтесь в основной поток", icon: direction };
  if (step.type === "off ramp") return { text: direction === "straight" ? "Двигайтесь к съезду" : `Съезд ${direction === "left" ? "налево" : "направо"}`, icon: direction };
  if (step.type === "on ramp") return { text: "Выезд на дорогу", icon: direction };
  if (modifier === "slight left" || modifier === "slight right") return { text: `Плавно ${side}`, icon: direction };
  if (modifier === "sharp left" || modifier === "sharp right") return { text: `Резко ${direction === "left" ? "налево" : "направо"}`, icon: direction };
  if (direction !== "straight") return { text: `Поверните ${direction === "left" ? "налево" : "направо"}`, icon: direction };
  return { text: "Продолжайте движение", icon: "straight" };
}

export function navigationDistance(value) {
  if (!numeric(value)) return "—";
  return value < 1000 ? `${Math.max(0, Math.round(value / 10) * 10)} м` : `${(value / 1000).toLocaleString("ru-RU", { maximumFractionDigits: 1 })} км`;
}
